backend/auth/role_map_repository.go

```diff
@@ -0,0 +1,222 @@
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type RoleMapRepository struct {
+	// Subrole map is optional, if not provided, subroles will be ignored. Roles received in token are checked only with RoleMap
+	RoleMap map[string]*models.Role
+	SubroleMap map[string]*models.Role
+}
+
+var (
+	instance      *RoleMapRepository
+	once          sync.Once
+)
+
+func GetInstance() (*RoleMapRepository, error) {
+	once.Do(func() {
+		roleMapNamespace := os.Getenv("ROLEMAP_NAMESPACE")
+		if roleMapNamespace == "" {
+			log.Printf("ROLEMAP_NAMESPACE not set, using default namespace")
+			roleMapNamespace = "default"
+		}
+
+		roleMapName := os.Getenv("ROLEMAP_NAME")
+		if roleMapName == "" {
+			log.Printf("ROLEMAP_NAME not set, using default name")
+			roleMapName = "role-mapper"
+		}
+
+		roleMap, subroleMap := GetRoleMapConfig(roleMapNamespace, roleMapName)
+		if roleMap == nil {
+			log.Printf("Failed to initialize RoleMapRepository")
+			return
+		}
+		instance = &RoleMapRepository{RoleMap: roleMap, SubroleMap: subroleMap}
+
+		log.Printf("RoleMapRepository initialized with %d roles %d subroles", len(instance.RoleMap), len(instance.SubroleMap))
+	})
+
+	if instance == nil {
+		return nil, errors.New("failed to initialize RoleMapRepository")
+	}
+
+	return instance, nil
+}
+
+func (rmr *RoleMapRepository) HasPermission(rolenames []string, operation *models.Operation) bool {
+	visited := make(map[string]interface{})
+	for _, role := range rolenames {
+		if rmr.hasPermission(role, operation, visited) {
+			return true
+		}
+	}
+	return false
+}
+
+func (rmr *RoleMapRepository) hasPermission(rolename string, operation *models.Operation, visited map[string]interface{}) bool {
+	role := rmr.RoleMap[rolename]
+	if role == nil {
+		return false
+	}
+
+	for _, deny := range role.Deny {
+		if deny.IsSuper(operation) {
+			return false
+		}
+	}
+
+	for _, permit := range role.Permit {
+		if permit.IsSuper(operation) {
+			return true
+		}
+	}
+
+	// Recursively check subroles, if any matches, return true
+	for _, subrole := range role.Subroles {
+		if _, exists := visited[subrole]; !exists {
+			if rmr.subHasPermission(subrole, operation, visited) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (rmr *RoleMapRepository) subHasPermission(rolename string, operation *models.Operation, visited map[string]interface{}) bool {
+	visited[rolename] = nil
+
+	role := rmr.SubroleMap[rolename]
+	if role == nil {
+		return false
+	}
+
+	for _, deny := range role.Deny {
+		if deny.IsSuper(operation) {
+			return false
+		}
+	}
+
+	for _, permit := range role.Permit {
+		if permit.IsSuper(operation) {
+			return true
+		}
+	}
+
+	// Recursively check subroles, if any matches, return true
+	for _, subrole := range role.Subroles {
+		if _, exists := visited[subrole]; !exists {
+			if rmr.subHasPermission(subrole, operation, visited) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
```

Comment:
A lot of repetition.

Maybe you can even combine them into one function? Something like this:
```
func (rmr *RoleMapRepository) checkPermission(rolename string, operation *models.Operation, visited map[string]struct{}, roleMap map[string]*Role) bool {
    if _, exists := visited[rolename]; exists {
        return false
    }
    visited[rolename] = struct{}{}

    role := roleMap[rolename]
    if role == nil {
        return false
    }

    for _, deny := range role.Deny {
        if deny.IsSuper(operation) {
            return false
        }
    }

    for _, permit := range role.Permit {
        if permit.IsSuper(operation) {
            return true
        }
    }

    // Rekurencyjnie sprawdzamy pod-role
    for _, subrole := range role.Subroles {
        if rmr.checkPermission(subrole, operation, visited, roleMap) {
            return true
        }
    }

    return false
}
```

Reply:
Please accept my refactor 🙏

Reply:
![cat like](https://github.com/user-attachments/assets/1a3ae2eb-eede-493c-b446-950e0b0c982c)